core/src/smartcontracts/isi/asset.rs

```diff
@@ -134,310 +148,205 @@ pub mod isi {
             {
-                wsv.new_tx_amounts
-                    .lock()
-                    .push(transfer_quantity.into_metric());
+                wsv.new_tx_amounts.lock().push(self.object.to_f64());
             }
 
             wsv.emit_events([
                 AssetEvent::Removed(AssetChanged {
                     asset_id: source_id,
-                    amount: transfer.object.into(),
+                    amount: self.object.into(),
                 }),
                 AssetEvent::Added(AssetChanged {
                     asset_id: destination_id,
-                    amount: transfer.object.into(),
+                    amount: self.object.into(),
                 }),
             ]);
 
             Ok(())
         }
     }
 
-    /// Trait for collecting fields associated with Iroha Special Instructions for assets:
-    /// - asset value type
-    /// - default asset value
-    trait AssetInstructionInfo {
-        const EXPECTED_VALUE_TYPE: AssetValueType;
-        const DEFAULT_ASSET_VALUE: Self;
-    }
-
-    macro_rules! impl_asset_instruction_info {
-        ($ty:ty, $expected_value_type:expr, $default_value:expr) => {
-            impl AssetInstructionInfo for $ty {
-                const EXPECTED_VALUE_TYPE: AssetValueType = $expected_value_type;
-                const DEFAULT_ASSET_VALUE: Self = $default_value;
+    /// Assert that asset type is Numeric and that it satisfy asset definition spec
+    pub(crate) fn assert_numeric_spec(
+        object: &Numeric,
+        asset_definition: &AssetDefinition,
+    ) -> Result<NumericSpec, Error> {
+        let object_spec = NumericSpec::fractional(object.scale());
+        let object_asset_value_type = AssetValueType::Numeric(object_spec);
+        let asset_definition_spec = match asset_definition.value_type {
+            AssetValueType::Numeric(spec) => spec,
+            other => {
+                return Err(TypeError::from(Mismatch {
+                    expected: other,
+                    actual: object_asset_value_type,
+                })
+                .into())
             }
         };
+        asset_definition_spec.check(object).map_err(|_| {
+            TypeError::from(Mismatch {
+                expected: AssetValueType::Numeric(asset_definition_spec),
+                actual: object_asset_value_type,
+            })
+        })?;
+        Ok(asset_definition_spec)
     }
 
-    impl_asset_instruction_info!(u32, AssetValueType::Quantity, 0_u32);
-    impl_asset_instruction_info!(u128, AssetValueType::BigQuantity, 0_u128);
-    impl_asset_instruction_info!(Fixed, AssetValueType::Fixed, Fixed::ZERO);
-
     /// Asserts that asset definition with [`definition_id`] has asset type [`expected_value_type`].
     pub(crate) fn assert_asset_type(
         definition_id: &AssetDefinitionId,
         wsv: &WorldStateView,
-        expected_value_type: AssetValueType,
+        expected_value_type: AssetValueTypeDiscriminant,
```

Comment:
```suggestion
        expected_value_type: AssetValueType,
```

shouldn't you also check that scale is correct here?